src/blister.c

```diff
@@ -236,20 +236,18 @@ Obj CopyBlist (
     Int                 mut )
 {
     Obj copy;
-    Obj tmp;
 
     if (!IS_MUTABLE_OBJ(list)) {
         return list;
     }
 
     copy = DoCopyBlist(list, mut);
+
     /* leave a forwarding pointer */
-    tmp = NEW_PLIST( T_PLIST, 2 );
-    SET_LEN_PLIST( tmp, 2 );
-    SET_ELM_PLIST( tmp, 1, CONST_ADDR_OBJ(list)[0] );
-    SET_ELM_PLIST( tmp, 2, copy );
-    ADDR_OBJ(list)[0] = tmp;
+    ADDR_OBJ(list)[0] = copy;
```

Comment:
Can you just give me a brief idea what's going on here? I'm not really clear what the code was doing before, but assume you understand.

Reply:
I mean, we might just have been making some weird list for no obvious reason?

Reply:
When we copy an object A, we must make sure that if we copy it multiple times, then we return the same "copied object" B each time (think of copying the list `[A,A]`, the result should be `[B,B]`, not `[B,C]`). To implement that, we make the copy B, then keep a forwarding pointer to the copy inside A; if asked to copy A again later during the same copy operation, we just return that forwarding pointer to B. Finally, when copying is done, restore the value overwritten by the forwarding pointer.

In this case (actually, in all?), the value overwritten is the first word / first slot of the blist. What the old code did (and the code for positional and component and data object still does) is to create a plist of length 2, which stores the old length value, plus the pointer to the copy B; then inside of A, we store a pointer to B.

However, this is not really necessary here: to restore the length of A, we can simply read it from the copy B, which after all has the same length! Thus, the code change avoid creating this temporary plist, and one indirection.

We can't do this for pos/comp/dat objs, as there the overwritten value is the type, and the type of a copy can differ (e.g. if the copy changes the mutability).

Reply:
thanks